### ISSUE TITLE
FedDC

### ISSUE THREAD

**Issue by jafermarq** [adap/flower#2040]
# FedDC

* **Title:** Federated Learning from Small Datasets
* **Venue:** ICLR 2023
* **Link to paper:** https://openreview.net/forum?id=hDDV1lsRV8

## Do you want to work on this baseline? 

> 🌻 Check everything about the `Summer of Reproducibility` on [`flower.dev/summer`](https://flower.dev/summer/)

> All available baselines are listed in the [Summer of Reproducibility Dashboard](https://github.com/orgs/adap/projects/6/) and also in the [GitHub Issues with the `summer-of-reproducibility` label](https://github.com/adap/flower/labels/summer-of-reproducibility). The content is the same.

> 📝 It is advised to complete these steps before your start working on your code. But if you can't wait to implement your baseline with Flower (we totally understand it 😄), please ensure you follow the steps on [how to contribute a new baseline](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline).

What follows are the steps 1 & 2 in the Summer of Reproducibility instructions.

### 1. Join the Summer of Reproducibility program
- [x] Join the [Flower Slack](https://flower.dev/join-slack/) and say _"hi! 👋"_ in the channel `#summer-of-reproducibility`.
- [x] Pick a baseline from our curated list **<---------------------------------------- [you are doing this now]**
### 2. Define the scope of your contribution
- [x] **What are you going to reproduce?** Add a comment to your issue and tell us about your plan regarding this baseline: what experiments from the paper are you reproducing?, for which datasets ? the more details you provide us with the better !
- [x] **Check if you are eligible for a reward**. 
    > As we have to comply with US/EU regulations, we have checked that individual contributors based on these countries or territories are eligible: Australia, Austria, Belgium, Bulgaria, Canada, Croatia, Cyprus, Czech Republic, Denmark, Estonia, Finland, France, Germany, Gibraltar, Greece, Hong Kong SAR China, Hungary, India, Ireland, Italy, Japan, Latvia, Liechtenstein, Lithuania, Luxembourg, Malta, Mexico, Netherlands, New Zealand, Norway, Poland, Portugal, Romania, Singapore, Slovakia, Slovenia, South Korea, Spain, Sweden, Switzerland, Thailand, United Arab emirates, United Kingdom United States.

    If where you are based is not on the list, please send us an email (`summer@flower.dev`) letting us know a bit about yourself (where are you currently based?, are you a university student? do you work at a public institution?). Please tell us the baselines you are interested in implementing (i.e. tell us your GitHub issue if you have crated one). We will reach back to you.
- [x] We will discuss with you about your contribution plan, if it sounds like a substantial enough contribution according to the Summer of Reproducibility rules (check our website [flower.dev/summer](https://flower.dev/summer/)), you'll get the OK from us to start working on your baseline!

## What happens next?

- [x] This item will be moved to the `In Progress` stage by a member of the Flower Team.

- [ ] Follow the [instructions for creating a new baseline](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline) which will guide you through the process step-by-step.

## Is something wrong or not clear ?
* Ask a question directly in your issue.
* Reach out to us via the [Flower Slack](https://flower.dev/join-slack/) and ask your question in the #summer-of-reproducibility channel
* Check all the details (including FAQ) in the Summer of Reproducibility website: flower.dev/summer

**Comment by urwithajit9** [comment]
Hi @jafermarq 

I am interested in this baseline and want to reproduce Table 1. The paper is newly published, so before making a final decision, I need to read it thoroughly. 
I will make my decision and update you here. (current comment is as placeholder because my 1st and 2nd choice is already has picked by others).

Ajit

**Comment by maanavgoyal** [comment]
Hi @jafermarq 
It appears to me that this baseline is still up for grabs, so here is my proposal.
#### Contribution Plan
- The experiments will involve comparing the performance of FedDC against FedAvg and the vanilla daisy chain baseline.
- I plan to reproduce (Table 1 and Table 2) proposed in the paper using the CIFAR-10, MRI, and Pneumonia datasets.

I can confirm that i reside in of the eligible territories listed above.

**Comment by negedng** [comment]
Noo, I'm late :D

Implementing this paper in Flower seems simple. You have to store the client models in the strategy.aggregate_fit() and override the configurate_fit() to take from the stored client weight list instead of the aggregated one.

As long as `b mod d = 0` (which is the case in the experiments), you can make the local epochs d treat every round as a chain-update and every b/d-th epoch as a special case


```
    def configure_fit(
        self, server_round: int, parameters: Parameters, client_manager: ClientManager
    ) -> List[Tuple[ClientProxy, FitIns]]:
        """Configure the next round of training."""
        config = {}
        if self.on_fit_config_fn is not None:
            # Custom fit config function provided
            config = self.on_fit_config_fn(server_round)

        # if server_round%b//d==0: 
        fit_ins = FitIns(parameters, config)                                        # Override this line
        # else 
        # Get from stored weights

        # Sample clients
        sample_size, min_num_clients = self.num_fit_clients(
            client_manager.num_available()
        )
        clients = client_manager.sample(
            num_clients=sample_size, min_num_clients=min_num_clients
        )

        # Return client/config pairs
        return [(client, fit_ins) for client in clients]
```

I was planning to propose doing Table 1 and Table 2 with all the other methods as they are already implemented in Flower. Writing a DataLoader for the MRI and Pneumonia datasets looks like the hardest part of this implementation to me:)

**Comment by jafermarq** [comment]
Hi @maanavgoyal , your contribution plan makes sense. Thank you also for confirming your eligibility also. One question before proceeding, are you confident you have access to the compute resources (i.e. dedicated GPUs -- free cloud options might not be sufficient for this type of work) to run the experiments you indicate ?

**Comment by jafermarq** [comment]
amazing! Then you are all set. I have ✅ all points in Step 1 & 2, I added you as the _assignee_ to this issue and moved it to _In Progress_ status. Please remember the Flower _Summer of Reproducibility_ ends at the end of September so it is advised to start as soon as you can. You can find the steps on how to begin with all-things code related by following the link in the _What happens next?_ section in the description of this issue. If you have any questions feel free to reach out to me on the Flower Slack or by asking your question in our #questions channel. You can also reach out to all the other contributors via the #summer-of-reproducibility channel in Slack.

Looking forward to seeing your `FedDC` implementation in action! 🚀

**Comment by jafermarq** [comment]
Hi @maanavgoyal,
This is just a gentle reminder that the Flower *Summer of Reproducibility* is ending at the end of the month. With just a little more than 3 weeks to go, we are excited to see quite a few baselines well ahead in the process with [their respective PRs close to ready.](https://github.com/adap/flower/pulls?q=is%3Apr+is%3Aopen+label%3Asummer-of-reproducibility) If your PR is already on the list, great !! Please make sure the PR is linked to this issue (you just need to copy the URL of this issue somewhere in the main message of your PR). Ping me when you'd like me to take a look.

Also, make sure you keep an eye:eyes: on the **#summer-of-reproducibility** channel in the Flower Slack. I’ll announce very soon a new (the third!) round of 1:1 ask-me-anything sessions to help _Summer of Reproducibility_ contributors like yourself to meet the deadline. Please consider booking a time slot if you want to chat with me about your baseline, potential issues you have making your code run, how to open a PR, doubts about what to include in your readme, how to use Hydra configs more effective, etc … all questions are welcome!!

**Comment by jl2272** [comment]
Hi @jafermarq 

Is this baseline still available? If possible I would like to reproduce Table 1 and 2. Thanks

**Comment by jafermarq** [comment]
Hi @jl2272 , yes as far as I know this baseline is available. We welcome your contribution!! But please bear in mind that the _Summer of Reproducibility_ initiative is over. Still, me an other members in the Flower team we are here to help. Ping me when you have a first implementation of `FedDC` ready for review! You can reach out to us also [via our Slack workspace](https://flower.ai/join-slack/)

**Comment by WilliamLindskog** [comment]
Hi @jl2272, @maanavgoyal, @negedng, @urwithajit9, 

Just checking in here.Is anyone of you still eager to complete this baseline? 

Best regards
William

**Comment by WilliamLindskog** [comment]
Closing this as this baseline request has gone stale.
